### PR TITLE
Firebase realtime database

### PR DESCRIPTION
**Pull request by giommok**
- Implemented Firebase Realtime Database
- Created a Database class that will work as an interface for all our interactions with the database
- Added tests for Database class
- Added a simple test for Account class testing everything is fine when no account is signed in
- Updated sign-in so it will be easy to implement the "choose username" feature on the first login.

### CODE REVIEW

File: app/src/main/java/com/github/bgabriel998/softwaredevproject/Button2Activity.java [hunk 1]
```diff
@@ -113,43 +115,64 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
         }
     }
 
-    private void firebaseAuthWithGoogle() {
-        AuthCredential credential = GoogleAuthProvider.getCredential(account.getGoogleIdToken(), null);
-        account.getFirebaseAuth().signInWithCredential(credential)
+    private void firebaseAuthWithGoogle(String idToken) {
```
Comment (bgabriel998) [inline app/src/main/java/com/github/bgabriel998/softwaredevproject/Button2Activity.java:118]: Please add some JavaDoc comments for the method

File: app/src/main/java/com/github/bgabriel998/softwaredevproject/Button2Activity.java [hunk 2]
```diff
@@ -60,37 +66,34 @@ protected void onCreate(Bundle savedInstanceState) {
...
-                updateUI();
+                signOut();
                 break;
         }
     }
 
     /* The following three methods are necessary in order to let the user sign in */
     /* This is the public method to call in order to authenticate the user */
-    public void googleSignIn() {
+    public void signIn() {
```
Comment (bgabriel998) [inline app/src/main/java/com/github/bgabriel998/softwaredevproject/Button2Activity.java:94]: JavaDoc comments needed

File: app/src/main/java/com/github/bgabriel998/softwaredevproject/Button2Activity.java [hunk 3]
```diff
@@ -113,43 +115,64 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
-                            //Snackbar.make(mBinding.mainLayout, "Authentication Failed.", Snackbar.LENGTH_SHORT).show();
                             updateUI();
                         }
 
-                        // ...
                     }
                 });
     }
 
+    private void registerUser(String username) {
```
Comment (bgabriel998) [inline app/src/main/java/com/github/bgabriel998/softwaredevproject/Button2Activity.java:139]: JavaDoc comments needed

File: app/src/main/java/com/github/bgabriel998/softwaredevproject/Button2Activity.java [hunk 4]
```diff
@@ -113,43 +115,64 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+        Database.setChild("users/" + account.getId(), Arrays.asList("email", "username"), Arrays.asList(account.getEmail(), username));
+    }
+
+    private void isUsernameUsed(String username) {
```
Comment (bgabriel998) [inline app/src/main/java/com/github/bgabriel998/softwaredevproject/Button2Activity.java:143]: JavaDoc comments needed

File: app/src/main/java/com/github/bgabriel998/softwaredevproject/Button2Activity.java [hunk 5]
```diff
@@ -113,43 +115,64 @@ public void onActivityResult(int requestCode, int resultCode, Intent data) {
...
+        Database.isPresent("users", "username", username, () -> usernameAlreadyPresent() , () -> registerUser(username));
+    }
 
+    private void usernameAlreadyPresent() {
```
Comment (bgabriel998) [inline app/src/main/java/com/github/bgabriel998/softwaredevproject/Button2Activity.java:147]: JavaDoc comments needed

File: app/src/main/java/com/github/giommok/softwaredevproject/Account.java [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+     */
+    public String getEmail();
+
+    /**
+     * Returns the ID of the user or null if no account is signed in
+     * @return
+     */
+    public String getId();
+
+    public Uri getPhotoUrl();
```
Comment (bgabriel998) [inline app/src/main/java/com/github/giommok/softwaredevproject/Account.java:48]: JavaDoc comments needed

**Review by bgabriel998 — COMMENTED** [review]
Some comments are missing, otherwise no problem with the code

**Review by bgabriel998 — APPROVED** [review]
Great !

**Review by aolsson711 — APPROVED** [review]
Looks good!

**Pull request merged**